fuse_core/src/transaction_deserializer.cpp

```diff
@@ -0,0 +1,109 @@
+ *  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ *  FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ *  COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ *  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ *  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ *  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ *  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+#include <fuse_core/transaction_deserializer.h>
+
+#include <fuse_core/serialization.h>
+#include <fuse_msgs/SerializedTransaction.h>
+
+#include <algorithm>
+#include <sstream>
+#include <vector>
+
+
+namespace fuse_core
+{
+
+void serializeTransaction(const fuse_core::Transaction& transaction, fuse_msgs::SerializedTransaction& msg)
+{
+  std::stringstream stream;
+  // Serialize graph into the stream
+  // The archive to not flushed until it goes out of scope
+  {
+    BinaryOutputArchive archive(stream);
+    transaction.serialize(archive);
+  }
+  // Copy the stream into the message's data[] variable
+  copyStream(stream, msg.data);
+}
+
+TransactionDeserializer::TransactionDeserializer() :
+  constraint_loader_("fuse_core", "fuse_core::Constraint"),
+  variable_loader_("fuse_core", "fuse_core::Variable")
+{
+  // Load all known plugin libraries
+  // I believe the library containing a given Variable or Constraint must be loaded in order to deserialize
+  // an object of that type. But I haven't actually tested that theory.
+  for (const auto& class_name : constraint_loader_.getDeclaredClasses())
+  {
+    constraint_loader_.loadLibraryForClass(class_name);
+  }
+  for (const auto& class_name : variable_loader_.getDeclaredClasses())
+  {
+    variable_loader_.loadLibraryForClass(class_name);
+  }
+}
+
+TransactionDeserializer::~TransactionDeserializer()
+{
+  for (const auto& class_name : variable_loader_.getDeclaredClasses())
+  {
+    variable_loader_.unloadLibraryForClass(class_name);
+  }
+  for (const auto& class_name : constraint_loader_.getDeclaredClasses())
+  {
+    constraint_loader_.unloadLibraryForClass(class_name);
+  }
+}
+
+fuse_core::Transaction TransactionDeserializer::deserialize(const fuse_msgs::SerializedTransaction::ConstPtr& msg)
+{
+  return deserialize(*msg);
+}
+
+fuse_core::Transaction TransactionDeserializer::deserialize(const fuse_msgs::SerializedTransaction& msg)
+{
+  // The Transaction object is not a plugin and has no derived types. That makes it much easier to use.
+  auto transaction = fuse_core::Transaction();
+  // Deserialize the message into the Variable. This will throw if something goes wrong in the deserialization.
+  std::stringstream stream;
+  std::copy(msg.data.begin(), msg.data.end(), std::ostream_iterator<char>(stream));
```

Comment:
Boost has everything.

I did a quick attempt, but I ran into type issues. The archive only accepts streams of type `char`. However the serialized message is defined as `uint8`/`unsigned char`. And fascinatingly, switching the message to `int8`/`signed char` also doesn't work because a `char` is not the same type as a `signed char`. I could switch the message buffer to `string`, but then I pretty sure it would print even if you use `rostopic echo --noarr`. I'll get this working eventually, just not today.